File: internal/burn/burn.go
// Package burn projects weekly and 5h burn rates from the rate-limit history log the statusline
// records (~/.claude/rate-limit-history.jsonl). Reset-robust by design: a DROP in % (or a change of
// reset window) is treated as a reset boundary, so a surprise Anthropic reset can never produce a
// phantom negative burn or a doom projection — it just restarts the clock. Flat / just-reset history
// yields no projection, not a guess. Fail-soft: Read never panics and distinguishes an UNREADABLE /
// garbled history (readable=false) from a merely absent / empty one (readable=true, empty slice).
package burn

import (
	"encoding/json"
	"os"
	"sort"
	"strings"

	"github.com/SeanLF/ccpool/internal/rb"
)

// Entry is one history sample (or an envelope row): the raw parsed object, numbers kept as
// json.Number for read rows and float64 for envelope-built rows. Read numeric fields via num/tof.
type Entry = map[string]any

// Noise guards. Weekly % is a coarse integer, so it wiggles +-1 between renders; without these a
// 1-point blip over two minutes extrapolates to an absurd rate.
const (
	dropReset = 5.0 // a wk fall bigger than this = a real reset, not noise
	minSpanH  = 2.0 // need this many hours of run before a slope is trustworthy
	minDelta  = 3.0 // and this much net climb (else it's within rounding noise)

	// Short-horizon (5h SESSION) least-squares fit knobs.
	recentSecs     = 1800.0 // trailing window to fit (30 min)
	recentMinSpanH = 0.08   // ~5 min of run before a slope is trustworthy
	recentMinDelta = 2.0    // and this much climb (past integer noise)

	keepSecs = 14 * 86400 // drop samples older than this on read

	// resetJitter is how far below the max a reset may sit and still count as the SAME window.
	// Anthropic's reported resets_at wobbles a few SECONDS between renders, so a strict max lands on
	// a 1-sample outlier and collapses the envelope (burn/runway silently die). A window is >= 5h
	// away from its neighbour, so 300s can't merge two real windows.
	resetJitter = 300.0
)

// Projection is the weekly burn result: avg %/hour over the current run and the derived time-to-cap,
// plus the run bounds Runway re-measures in working hours.
type Projection struct {
	BurnPerH   float64
	HoursToCap float64
	Wk         float64
	FirstT     int64
	LastT      int64
	Dpct       float64
}

// Recent is the short-horizon (5h) fit: %/hour, time-to-cap, and the last value.
type Recent struct {
	RatePerH   float64
	HoursToCap float64
	Val        float64
}

// num reads a value as a float64, reporting whether it was numeric (mirrors Ruby is_a?(Numeric)).
// Handles both json.Number (read rows) and float64 (envelope-built rows).
func num(v any) (float64, bool) {
	switch n := v.(type) {
	case json.Number:
		f, err := n.Float64()
		return f, err == nil
	case float64:
		return n, true
	case int64:
		return float64(n), true
	case int:
		return float64(n), true
	}
	return 0, false
}

// tof mirrors Ruby's numeric .to_f: the value if numeric, else 0.0 (nil.to_f == 0.0).
func tof(v any) float64 { f, _ := num(v); return f }

// Read returns recent, well-formed samples (chronological). readable=false is the Ruby `nil`: the
// file exists with content but nothing parses (drift/corruption) or it can't be read at all. An
// absent or empty file returns (empty, true). Never panics.
func Read(path string, now int64) (entries []Entry, readable bool) {
	b, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return []Entry{}, true // absent -> []
		}
		return nil, false // I/O error (EACCES, EISDIR, ...) -> unreadable
	}

	var parsed []Entry
	anyContent := false
	for _, l := range strings.Split(string(b), "\n") {
		if strings.TrimSpace(l) != "" {
			anyContent = true
		}
		if e := parse(l); e != nil {
			parsed = append(parsed, e)
		}
	}
	// Content present but not a single usable line -> drift/corruption, not warmup.
	if len(parsed) == 0 && anyContent {
		return nil, false
	}

	cutoff := now - keepSecs
	kept := []Entry{} // non-nil, matching the absent-file branch's ([]Entry{}, true) contract
	for _, e := range parsed {
		if t, ok := num(e["t"]); ok && int64(t) >= cutoff {
			kept = append(kept, e)
		}
	}
	sort.SliceStable(kept, func(i, j int) bool { return tof(kept[i]["t"]) < tof(kept[j]["t"]) })
	return kept, true
}

// parse validates one line: a JSON object with numeric t and wk. A bad line is dropped (nil), not
// fatal; Read flags an all-bad, non-empty file as unreadable.
func parse(line string) Entry {
	e := rb.ParseObject([]byte(line))
	if e == nil {
		return nil
	}
	if _, ok := num(e["t"]); !ok {
		return nil
	}
	if _, ok := num(e["wk"]); !ok {
		return nil
	}
	return e
}

// Envelope collapses the raw multi-session log into ONE account-global monotonic series for a single
// field over its CURRENT window: the running MAX over the latest reset window (older windows are
// prior weeks/sessions and are dropped). Returns nil when entries is nil, so an unreadable history
// stays distinguishable from an empty one.
func Envelope(entries []Entry, field, resetField string) []Entry {
	if entries == nil {
		return nil
	}
	// latest reset among rows where both field and reset are numeric (non-numeric resets can't be
	// ordered and are excluded; the loop's `== latest` then drops them too).
	var latest float64
	hasLatest := false
	for _, e := range entries {
		_, fok := num(e[field])
		rv, rok := num(e[resetField])
		if fok && rok {
			if !hasLatest || rv > latest {
				latest = rv
			}
			hasLatest = true
		}
	}

	var out []Entry
	var running float64
	haveRunning := false
	for _, e := range entries {
		fv, fok := num(e[field])
		if !fok {
			continue
		}
		if hasLatest {
			rv, rok := num(e[resetField])
			if !rok || latest-rv > resetJitter {
				continue // current window only (bucketing jittered resets within tolerance of the max)
			}
		} else if e[resetField] != nil {
			continue // no reset recorded anywhere: only rows with a nil reset pass (nil == nil)
		}
		if !haveRunning || fv > running {
			running = fv
		}
		haveRunning = true
		row := Entry{"t": e["t"], field: running}
		if hasLatest {
			row[resetField] = latest
		}
		out = append(out, row)
	}
	return out
}

// currentRun is the longest trailing run since the most recent reset boundary. A reset is a BIG drop
// (> dropReset) or a change of reset window; small -1 dips stay inside the run.
func currentRun(entries []Entry) []Entry {
	if len(entries) < 2 {
		return entries
	}
	start := len(entries) - 1
	for i := len(entries) - 1; i >= 1; i-- {
		if tof(entries[i]["wk"]) < tof(entries[i-1]["wk"])-dropReset {
			break // weekly fell hard -> a reset landed here
		}
		if resetDiffers(entries[i]["wk_reset"], entries[i-1]["wk_reset"]) {
			break // window changed -> new week
		}
		start = i - 1
	}
	return entries[start:]
}

// resetDiffers mirrors Ruby `a != b` for reset values: both nil -> equal, numeric-equal -> equal.
func resetDiffers(a, b any) bool {
	av, aok := num(a)
	bv, bok := num(b)
	if aok != bok {
		return true
	}
	if aok {
		return av != bv
	}
	return false
}

// Project is the average %/hour over the current run and the resulting time-to-cap. ok=false when
// there isn't enough signal (no entries, flat, or just reset).
func Project(entries []Entry) (Projection, bool) {
	if len(entries) < 2 {
		return Projection{}, false
	}
	run := currentRun(entries)
	if len(run) < 2 {
		return Projection{}, false
	}
	first, last := run[0], run[len(run)-1]
	ft, lt := tof(first["t"]), tof(last["t"])
	dtH := (lt - ft) / 3600.0
	dpct := tof(last["wk"]) - tof(first["wk"])
	// Need a real climb over a real span: too little of either and the slope is noise.
	if dtH < minSpanH || dpct < minDelta {
		return Projection{}, false
	}
	burn := dpct / dtH
	return Projection{
		BurnPerH:   burn,
		HoursToCap: (100.0 - tof(last["wk"])) / burn,
		Wk:         tof(last["wk"]),
		FirstT:     int64(ft),
		LastT:      int64(lt),
		Dpct:       dpct,
	}, true
}

// trailingRun is the trailing run of a field since its last reset (a drop > dropReset). Detects a
// session reset purely from a drop in value (no reset-window column for ses).
func trailingRun(entries []Entry, field string) []Entry {
	if len(entries) < 2 {
		return entries
	}
	start := len(entries) - 1
	for i := len(entries) - 1; i >= 1; i-- {
		if tof(entries[i][field]) < tof(entries[i-1][field])-dropReset {
			break
		}
		start = i - 1
	}
	return entries[start:]
}

// lsqSlope is the ordinary least-squares slope (field-% per hour) over a run. ok=false when the
// x-variance is zero.
func lsqSlope(run []Entry, field string) (float64, bool) {
	n := float64(len(run))
	t0 := tof(run[0]["t"])
	xs := make([]float64, len(run))
	ys := make([]float64, len(run))
	var sx, sy float64
	for i, e := range run {
		xs[i] = (tof(e["t"]) - t0) / 3600.0
		ys[i] = tof(e[field])
		sx += xs[i]
		sy += ys[i]
	}
	mx, my := sx/n, sy/n
	var den float64
	for _, x := range xs {
		den += (x - mx) * (x - mx)
	}
	if den == 0 {
		return 0, false
	}
	var numr float64
	for i := range xs {
		numr += (xs[i] - mx) * (ys[i] - my)
	}
	return numr / den, true
}

// ProjectRecent is the short-horizon projection for a fast window: fit the trailing recentSecs of
// samples carrying field and extrapolate to 100%. ok=false when there's no real, positive climb over
// a real span.
func ProjectRecent(entries []Entry, now int64, field string) (Recent, bool) {
	if len(entries) == 0 {
		return Recent{}, false
	}
	var recent []Entry
	for _, e := range entries {
		if t, ok := num(e["t"]); ok && int64(t) >= now-int64(recentSecs) {
			if _, ok := num(e[field]); ok {
				recent = append(recent, e)
			}
		}
	}
	if len(recent) < 3 {
		return Recent{}, false
	}
	run := trailingRun(recent, field)
	if len(run) < 3 {
		return Recent{}, false
	}
	spanH := (tof(run[len(run)-1]["t"]) - tof(run[0]["t"])) / 3600.0
	delta := tof(run[len(run)-1][field]) - tof(run[0][field])
	if spanH < recentMinSpanH || delta < recentMinDelta {
		return Recent{}, false
	}
	slope, ok := lsqSlope(run, field)
	if !ok || slope <= 0 {
		return Recent{}, false
	}
	last := tof(run[len(run)-1][field])
	return Recent{RatePerH: slope, HoursToCap: (100.0 - last) / slope, Val: last}, true
}

File: burn.rb
# frozen_string_literal: true

# Weekly burn-rate projection from the rate-limit history the status line records
# (~/.claude/rate-limit-history.jsonl, one JSON sample per change).
#
# Reset-robust by design. Anthropic resets the weekly window WITHOUT warning --
# model launches, rate-limit-counting fixes, incidents -- so resets_at is not
# trustworthy as the only reset signal. Instead we treat any DROP in weekly %
# (or a change of reset window) as a reset boundary and measure burn only over
# the current monotonic run since the last boundary. A surprise reset can never
# produce a phantom "negative burn" or a doom projection: it just restarts the
# clock. Flat or just-reset history yields nil (no burn to project), not a guess.
#
# Fail-soft but not fail-silent: read() never raises, and it distinguishes an
# UNREADABLE/garbled history (returns nil) from a merely absent/empty one
# (returns []). The caller surfaces the former instead of quietly acting as if
# there's simply no burn -- a dead projection must not masquerade as "all clear".
require "json"

module Burn
  module_function

  # Noise guards. Weekly % is a coarse integer, so it wiggles +-1 between renders;
  # without these a 1-point blip over two minutes extrapolates to an absurd rate.
  DROP_RESET = (ENV["USAGE_BURN_DROP_RESET"] || "5").to_f   # a wk fall bigger than this = a real reset, not noise
  MIN_SPAN_H = (ENV["USAGE_BURN_MIN_SPAN_H"] || "2").to_f   # need this many hours of run before a slope is trustworthy
  MIN_DELTA  = (ENV["USAGE_BURN_MIN_DELTA"]  || "3").to_f   # and this much net climb (else it's within rounding noise)

  # Short-horizon (5h SESSION) rate. The weekly two-point slope is too coarse for a
  # window that moves in minutes, so we fit a line by least-squares over a trailing
  # window: every sample pins the fit, so a lone +-1 integer wiggle can't swing it.
  RECENT_SECS       = (ENV["USAGE_SES_WINDOW_SECS"] || "1800").to_f # trailing window to fit (30 min)
  RECENT_MIN_SPAN_H = (ENV["USAGE_SES_MIN_SPAN_H"]  || "0.08").to_f # ~5 min of run before a slope is trustworthy
  RECENT_MIN_DELTA  = (ENV["USAGE_SES_MIN_DELTA"]   || "2").to_f    # and this much climb (past integer noise)

  # => Array of recent, well-formed samples (chronological); [] when there's no
  # history yet; nil when the file exists with content but nothing parses (schema
  # drift / corruption) or can't be read at all. Never raises.
  def read(path, now, keep_secs: 14 * 86_400)
    return [] unless File.exist?(path)

    raw = File.readlines(path)
    parsed = raw.filter_map { |l| parse(l) }
    # Content present but not a single usable line -> drift/corruption, not warmup.
    return nil if parsed.empty? && raw.any? { |l| !l.strip.empty? }

    parsed.select { |e| e["t"] >= now - keep_secs }.sort_by { |e| e["t"] }
  rescue StandardError
    nil  # I/O error (EACCES, EISDIR, bad encoding...) -> unreadable, signal it
  end

  def parse(line)
    e = JSON.parse(line)
    e if e.is_a?(Hash) && e["t"].is_a?(Numeric) && e["wk"].is_a?(Numeric)
  rescue StandardError
    nil  # a bad line is dropped, not fatal; read() flags an all-bad file as nil
  end

  # Collapse the raw multi-session log into ONE account-global monotonic series for a
  # single field over its CURRENT window. The status line records every concurrent
  # session's samples into one file, each frozen at a different staleness, so a naive
  # read interleaves them (wk 42, 35, 42, 32...) and current_run/trailing_run see
  # phantom resets at every downward step. Both windows are account-global and
  # monotonic within a window, so the truth is the running MAX over the current window
  # (the latest reset_field value); samples from older windows are prior weeks/sessions
  # and are dropped. => chronological Array<{t, field[, reset_field]}> (possibly empty),
  # or nil when entries is nil -- an unreadable history must stay distinguishable from
  # an empty one, not be masked as "no burn".
  def envelope(entries, field, reset_field)
    return nil if entries.nil?

    # reset_field must be Numeric to be ordered: parse() only validates t and the
    # field itself, so a schema-drifted line can carry a non-numeric reset. Ordering
    # a String against an Integer would raise -- fatal here, since the caller invokes
    # envelope outside any rescue. Non-numeric resets are simply excluded (the loop's
    # `== latest` then drops them too), preserving the nil==nil legacy-fallback path.
    latest  = entries.filter_map { |e| e[reset_field] if e[field].is_a?(Numeric) && e[reset_field].is_a?(Numeric) }.max
    running = nil
    entries.filter_map do |e|
      next unless e[field].is_a?(Numeric)
      next unless e[reset_field] == latest  # current window only (nil == nil when no reset recorded)

      running = running ? [running, e[field]].max : e[field]
      row = { "t" => e["t"], field => running }
      row[reset_field] = latest if latest
      row
    end
  end

  # The longest trailing run since the most recent reset boundary. A reset is a
  # BIG drop (> DROP_RESET) or a change of reset window -- small -1 dips are
  # rounding noise and stay inside the run (netted out by first->last), so a blip
  # can't fragment the history into a tiny, over-steep run.
  def current_run(entries)
    return entries if entries.size < 2

    start = entries.size - 1
    (entries.size - 1).downto(1) do |i|
      cur = entries[i]
      prev = entries[i - 1]
      break if cur["wk"] < prev["wk"] - DROP_RESET   # weekly fell hard -> a reset landed here
      break if cur["wk_reset"] != prev["wk_reset"]   # window changed -> new week

      start = i - 1
    end
    entries[start..]
  end

  # Average %/hour over the current run and the resulting time-to-cap.
  # `entries` comes from read(). => { burn_per_h:, hours_to_cap:, wk: } or nil
  # when there isn't enough signal (no entries, flat, or just reset).
  def project(entries)
    return nil unless entries && entries.size >= 2

    run = current_run(entries)
    return nil if run.size < 2

    first = run.first
    last = run.last
    dt_h = (last["t"] - first["t"]) / 3600.0
    dpct = last["wk"] - first["wk"]
    # Need a real climb over a real span: too little of either and the slope is
    # noise, not a trend. Better to project nothing than to cry wolf.
    return nil if dt_h < MIN_SPAN_H || dpct < MIN_DELTA

    burn = dpct / dt_h
    # first_t/last_t/dpct let Runway re-measure the same run in WORKING hours (∫ Profile weight)
    # rather than the wall-clock hours this slope is over.
    { burn_per_h: burn, hours_to_cap: (100.0 - last["wk"]) / burn, wk: last["wk"],
      first_t: first["t"], last_t: last["t"], dpct: dpct }
  end

  # The trailing run of a field since its last reset (a drop > DROP_RESET = the
  # window rolled over). Parameterized by field so it works for "ses" as well as
  # "wk"; unlike current_run it has no reset-window column to lean on (history only
  # records wk_reset), so it detects a session reset purely from a drop in value.
  def trailing_run(entries, field)
    return entries if entries.size < 2

    start = entries.size - 1
    (entries.size - 1).downto(1) do |i|
      break if entries[i][field].to_f < entries[i - 1][field].to_f - DROP_RESET
      start = i - 1
    end
    entries[start..]
  end

  # Ordinary least-squares slope (units of field-% per hour) over a run. Robust to
  # the integer wiggle that makes a two-point slope unreliable on a short span.
  def lsq_slope(run, field)
    n  = run.size.to_f
    t0 = run.first["t"]
    xs = run.map { |e| (e["t"] - t0) / 3600.0 } # hours since run start
    ys = run.map { |e| e[field].to_f }
    mx = xs.sum / n
    my = ys.sum / n
    den = xs.sum { |x| (x - mx)**2 }
    return nil if den.zero?

    xs.zip(ys).sum { |x, y| (x - mx) * (y - my) } / den
  end

  # Short-horizon projection for a fast window. Fits the trailing `window_secs` of
  # samples carrying `field` and extrapolates to 100%.
  # => { rate_per_h:, hours_to_cap:, val: } or nil when there's no real, positive
  # climb over a real span (idle, flat, just-reset, or too few points to trust).
  def project_recent(entries, now, field:, window_secs: RECENT_SECS)
    return nil unless entries && !entries.empty?

    recent = entries.select { |e| e["t"] >= now - window_secs && e[field].is_a?(Numeric) }
    return nil if recent.size < 3

    run = trailing_run(recent, field)
    return nil if run.size < 3

    span_h = (run.last["t"] - run.first["t"]) / 3600.0
    delta  = run.last[field].to_f - run.first[field].to_f
    return nil if span_h < RECENT_MIN_SPAN_H || delta < RECENT_MIN_DELTA

    slope = lsq_slope(run, field)
    return nil unless slope&.positive?

    { rate_per_h: slope, hours_to_cap: (100.0 - run.last[field].to_f) / slope, val: run.last[field].to_f }
  end
end
